pennylane/pauli/pauli_arithmetic.py

```diff
@@ -0,0 +1,261 @@
+sparse_matZ = sparse.csr_matrix([[1, 0], [0, -1]])
+
+sparse_mat_map = {
+    I: sparse_matI,
+    X: sparse_matX,
+    Y: sparse_matY,
+    Z: sparse_matZ,
+}
+
+_map_I = {
+    I: (1, I),
+    X: (1, X),
+    Y: (1, Y),
+    Z: (1, Z),
+}
+_map_X = {
+    I: (1, X),
+    X: (1, I),
+    Y: (1.0j, Z),
+    Z: (-1.0j, Y),
+}
+_map_Y = {
+    I: (1, Y),
+    X: (-1.0j, Z),
+    Y: (1, I),
+    Z: (1j, X),
+}
+_map_Z = {
+    I: (1, Z),
+    X: (1j, Y),
+    Y: (-1.0j, X),
+    Z: (1, I),
+}
+
+mul_map = {I: _map_I, X: _map_X, Y: _map_Y, Z: _map_Z}
+
+
+class PauliWord(dict):
+    """Immutable dictionary used to represent a Pauli Word.
+    Can be constructed from a standard dictionary.
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: Z})
+    """
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping):
+        """Strip identities from PauliWord on init!"""
+        mapping = {wire: op for wire, op in mapping.items() if op != I}
+        super().__init__(mapping)
+
+    def __copy__(self):
+        """Copy the PauliWord instance."""
+        return PauliWord(dict(self.items()))
+
+    def __setitem__(self, key, item):
+        """Restrict setting items after instantiation."""
+        raise NotImplementedError
+
+    def update(self, __m, **kwargs) -> None:
+        """Restrict updating PW after instantiation."""
+        raise NotImplementedError
+
+    def __hash__(self):
+        return hash(frozenset(self.items()))
+
+    def __mul__(self, other):
+        """Multiply two Pauli words together using the matrix product if wires overlap
+        and the tensor product otherwise.
+
+        Args:
+            other (PauliWord): The Pauli word to multiply with
+
+        Returns:
+            result(PauliWord): The resulting operator of the multiplication
+            coeff(complex): The complex phase factor
+        """
+        c_self, c_other = (copy(self), copy(other))
+        result, iterator, swapped = (
+            (dict(c_self), c_other, False)
+            if len(self) > len(other)
+            else (dict(c_other), c_self, True)
+        )
+        coeff = 1
+
+        for wire, term in iterator.items():
+            if wire in result:
+                factor, new_op = (
+                    mul_map[term][result[wire]] if swapped else mul_map[result[wire]][term]
+                )
+                if new_op == I:
+                    del result[wire]
+                else:
+                    coeff *= factor
+                    result[wire] = new_op
+            elif term != I:
+                result[wire] = term
+
+        return PauliWord(result), coeff
+
+    def __str__(self):
+        """String representation of a PauliWord."""
+        if len(self) == 0:
+            return "[()]"
+        return " @ ".join(f"[{op}({w})]" for w, op in self.items())
```

Comment:
Why did you add square brackets around each Pauli op? Why not:
```pycon
>>> print(PauliWord({0: "X", 1: "Y"}))
X(0) @ Y(1)
```

Reply:
Just a stylistic choice, updating now